packages/rOpenScPCA/R/evaluate-clusters.R

```diff
@@ -108,3 +108,139 @@ calculate_purity <- function(
+#'   SingleCellExperiment objects, and "pca" for Seurat objects.
+#' @param ... Additional arguments to pass to `calculate_clusters()` which calculates
+#'   bootstrapped clusters. Usually, these will be the same arguments used to generate
+#'   the original clusters.
+#'
+#' @return Data frame with columns `replicate` and `ari`, representing the given bootstrap replicate
+#'   and its ARI value, respectively, and columns representing clustering algorithm parameters which
+#'   include at least `algorithm`, `weighting`, and `nn`. Louvain and leiden clustering will also
+#'   include `resolution`, and leiden clustering will further include `objective_function`.
+#'
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#'
+#' # First, cluster PCs from a SingleCellExperiment object using default parameters
+#' # and setting a seed for reproducibility
+#' cluster_df <- calculate_clusters(sce_object, seed = 11)
+#' # Second, calculate cluster stability using default parameters
+#' stability_df <- calculate_stability(sce_object, cluster_df$clusters, seed = 11)
+#'
+#'
+#' # First, cluster PCs from a SingleCellExperiment object using default parameters
+#' # and setting a seed for reproducibility
+#' cluster_df <- calculate_clusters(sce_object, seed = 11)
+#' # Second, calculate cluster stability using default parameters and 50 replicates
+#' stability_df <- calculate_stability(
+#'   sce_object,
+#'   cluster_df$clusters,
+#'   replicates = 50,
+#'   seed = 11
+#' )
+#'
+#'
+#' # First, cluster PCs from a SingleCellExperiment object using the leiden
+#' # algorithm and a resolution of 0.1
+#' cluster_df <- calculate_clusters(
+#'   sce_object,
+#'   algorithm = "leiden",
+#'   resolution = 0.1,
+#'   seed = 11
+#' )
+#' # Second, calculate cluster stability using the same parameters as were used
+#' # for the initial clustering
+#' stability_df <- calculate_stability(
+#'   sce_object,
+#'   cluster_df$clusters,
+#'   algorithm = "leiden",
+#'   resolution = 0.1,
+#'   seed = 11
+#' )
+#' }
+calculate_stability <- function(
+    x,
+    clusters,
+    replicates = 20,
+    seed = NULL,
+    pc_name = NULL,
+    ...) {
+  if (!is.null(seed)) {
+    set.seed(seed)
+  }
+
+  # ensure we have a matrix
+  pca_matrix <- prepare_pc_matrix(x, pc_name = pc_name)
+
+  # check clusters and matrix compatibility
+  stopifnot(
+    "The number of rows in the matrix must equal the length of the clusters vector." =
+      nrow(pca_matrix) == length(clusters)
+  )
+
+  # calculate ARI for each cluster result bootstrap replicate
+  all_ari_df <- 1:replicates |>
+    purrr::map(
+      \(i) {
+        sample_cells <- sample(nrow(pca_matrix), replace = TRUE)
+        resampled_pca <- pca_matrix[sample_cells, ]
+        original_clusters <- clusters[sample_cells]
+
+        resampled_df <- calculate_clusters(resampled_pca, ...)
+
+        ari <- pdfCluster::adj.rand.index(resampled_df$cluster, clusters[sample_cells])
```

Comment:
Since you went to the effort to define it, may as well use it!
```suggestion
        ari <- pdfCluster::adj.rand.index(resampled_df$cluster, original_clusters)
```